### PR TITLE
fix(react-router): support React Router 6 style relative paths in IonRouterOutlet

### PR DESCRIPTION
**Pull request by ShaneK**
Issue number: resolves [an issue from a comment](https://github.com/ionic-team/ionic-framework/issues/24177#issuecomment-3624311206)

---------

## What is the current behavior?

Currently, Ionic's RR6 doesn't support relative routes in the same way RR6 does. Routes that do not start with a `/`do not work in the Ionic RR6 implementation in some cases.

## What is the new behavior?

With this change, we properly support these route styles and more closely align with normal RR6 route support.

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

## Other information

Current dev build:
```
 8.7.12-dev.11765307927.1f491e92
```

This PR will be merged into the RR6 branch, which will be squash+merged into the major 9 branch. This will prevent major 9 from having commits in the change log on release that reference fixing things that are only available in major 9.

### CODE REVIEW

**Comment by ptmkenny** [comment]
I think there is a minor documentation issue here: `Routes that start with a / do not work in the Ionic RR6 implementation.` should be `Routes that **do not** start with a / do not work in the Ionic RR6 implementation.`

**Comment by ptmkenny** [comment]
I just installed the test branch for this:

```
 npm install @ionic/core@8.7.12-dev.11765219790.17cbe2e9 @ionic/react@8.7.12-dev.11765219790.17cbe2e9 @ionic/react-router@8.7.12-dev.11765219790.17cbe2e9
```

Unfortunately, this immediately breaks my app. `npm run dev` shows this error on the index page:

```
A <Route> is only ever to be used as the child of <Routes> element, never rendered directly. Please wrap your <Route> in a <Routes>.
```

I have all my `Route` components wrapped in `IonRouterOutlet`, not `Routes` (to match the structure of the code used in the tests). This was working with the other build, but in this build, I get the above error.

Changing all `IonRouterOutlet` components to `Routes` components works-- I can then navigate around my app normally.

**Comment by ShaneK** [comment]
@ptmkenny Can you provide a reproduction of this? I can't see how this change would cause that and I can't see how the e2e tests would be passing (much less building) if this were happening. [For example, this newly added one in this PR, which uses `IonRouterOutlet` + `Route`](https://github.com/ionic-team/ionic-framework/pull/30844/files#diff-e6d2a996a02d93eae4481eae9ef51be56fcd3a2219fadbbe3812428d37ac5706)

**Comment by ptmkenny** [comment]
@ShaneK Sorry about that, it seems there was some weird caching issue that caused that error.  I rebooted the machine, blew away all the caches, and I could build my app again with the new build.

However, it seems that not having a slash in a root-level route will still break the app. See the repro here: https://github.com/ptmkenny/ionic-react-router-6-test/tree/router-outlet

I put the reproduction instructions in the README.md. Also, note that in addition to adding the slash (for IonRouterOutlet), you can also replace IonRouterOutlet with Routes (without adding the slash) and the page will be shown.

**Comment by ShaneK** [comment]
@ptmkenny Wow, that was a big one! I don't know how I didn't have that as a test case before, but I do now.

I have a dev build up that will hopefully work for you now, as well as having this case as [part of my E2E tests](https://github.com/ionic-team/ionic-framework/pull/30844/files#diff-dc711e1dd902302b5e24a10271a2fc7b5440b56684c4f94b05f18fab80df9de5):
```
 8.7.12-dev.11765307927.1f491e92
 ```

**Comment by ptmkenny** [comment]
@ShaneK Thank you for the speedy fix, but it seems to be a little more complicated. 

Follow-up: https://github.com/ionic-team/ionic-framework/issues/24177#issuecomment-3637401533

**Comment by ShaneK** [comment]
That would be a new issue, not the same one. You should report it in the main thread and I'll deal with it in another PR

File: packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  IonBackButton,
+  IonButtons,
+} from '@ionic/react';
+import React from 'react';
+import { Route } from 'react-router-dom';
+
+/**
+ * This test page verifies that IonRouterOutlet correctly handles
+ * relative paths (paths without a leading slash) the same way
+ * React Router 6's Routes component does.
```
Comment (brandyscarney) [inline packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx:20]:
```suggestion
 * React Router 5's Routes component does.
```

Shouldn't this be a 5?

Do we even need to mention RR5?

File: packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ * Issue: https://github.com/ionic-team/ionic-framework/issues/24177#issuecomment-3624311206
```
Comment (brandyscarney) [inline packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx]:
```suggestion
```

I understand linking issues for regressions but I don't think we need to link to issue comments for basic functionality that existed in rr5.

File: packages/react-router/test/base/tests/e2e/specs/relative-paths.cy.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+  it('should navigate to Page B (defined with relative path - no leading slash)', () => {
+    cy.visit(`http://localhost:${port}/relative-paths`);
+    cy.ionPageVisible('relative-paths-home');
+    cy.ionNav('ion-item', 'Go to Page B');
+    cy.ionPageVisible('relative-paths-page-b');
+    cy.get('[data-testid="page-b-content"]').should('contain', 'Page B');
+  });
+
+  it('should navigate to Page C (defined with relative path - no leading slash)', () => {
```
Comment (brandyscarney) [inline packages/react-router/test/base/tests/e2e/specs/relative-paths.cy.js]: What's the point of a 2nd test for a relative path?

**Review by brandyscarney — APPROVED** [review]
LGTM! 🔥

File: packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx (hunk 1 above)
Comment (ShaneK) [reply packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx:20]: Nope, this functionality did not exist in RR5. RR5 used all absolute links. This is intentionally referring to RR6.

File: packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx (hunk 2 above)
Comment (ShaneK) [reply packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx]: This didn't exist in RR5

File: packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx (hunk 1 above)
Comment (brandyscarney) [reply packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx:20]: Ahhh sorry I was reading it like it was referring to our own router 😮‍💨

File: packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx (hunk 2 above)
Comment (brandyscarney) [reply packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx]: Still - this is basic functionality that is provided by RR6 then.

Comment (ShaneK) [reply packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx]: Yeah, and we need to mimic it. I guess it doesn't need to be linked in a comment like this, but we do it with other regressions we find so 🤷‍♂️

Comment (brandyscarney) [reply packages/react-router/test/base/src/pages/relative-paths/RelativePaths.tsx]: I think it would be more informative to link to the RR6 docs if we are just trying to make sure we support something. It's not a regression like "relative routing doesn't work in modals" it's an entire base concept that needs to be supported.

File: packages/react-router/test/base/tests/e2e/specs/relative-paths.cy.js (hunk 3 above)
Comment (ShaneK) [reply packages/react-router/test/base/tests/e2e/specs/relative-paths.cy.js]: I thought I was testing another edge case, but it seems not. Probably something I ended up forgetting to clean up. Good catch! [f570837](https://github.com/ionic-team/ionic-framework/pull/30844/commits/f570837a6ede64b91860debf6ab699d2d6521815)

File: packages/react-router/src/ReactRouter/ReactRouterViewStack.tsx [hunk 4]
```diff
@@ -469,7 +469,9 @@ export class ReactRouterViewStack extends ViewStacks {
     let parentPath: string | undefined = undefined;
     try {
       // Only attempt parent path computation for non-root outlets
-      if (outletId !== 'routerOutlet') {
+      // Root outlets have IDs like 'routerOutlet' or 'routerOutlet-2'
+      const isRootOutlet = outletId.startsWith('routerOutlet');
```
Comment (thetaPC) [inline packages/react-router/src/ReactRouter/ReactRouterViewStack.tsx:481]: Since this is being used in several files, why not create a utility for it?

Comment (ShaneK) [reply packages/react-router/src/ReactRouter/ReactRouterViewStack.tsx:481]: I mean, I haven't up until this point because it's such a simple single line of code, but I do agree that it probably should still be extracted into a utility - especially because it's not impossible that it changes at some point in the future. I think most of the refactoring I'll save until I have the RR6 branch more consolidated, I still have stuff pretty spread out and I'm not even sure how similar this file is in my next PR. But I think after the main RR6 branch is approved and all of the sub branches are merged into it, a single refactoring sweep would be a good idea to help clean up things like this and extremely large code blocks.

File: packages/react-router/test/base/tests/e2e/specs/relative-paths.cy.js (hunk 3 above)
Comment (brandyscarney) [reply packages/react-router/test/base/tests/e2e/specs/relative-paths.cy.js]: Maybe you meant to navigate from one relative path to another? 🤔
